### PR TITLE
Switch eventemitter2 to local asynceventemitter

### LINKED ISSUE #2366: EventEmitter2 changes how async events work
In old `async-eventemitter`:

`.on('step', (e) => {})` -> immediately resolves
`.on('step', (e, fn) => {setInterval(fn,1000)}` -> hangs interpreter for 1 s
`.on('step', async (e, fn) => {setInterval(fn,1000)}` -> hangs interpreter for 1 s
`.on('step', (e, fn) => {}` -> quits js (?) (I guess this is because node detects that there are only hanging promises left, and nothing is ran, so it deadlocks and therefore just quits)

In `EventEmitter2`:

```
.on('step') (e: InterpreterStep) => {
return new Promise((resolve) => {setTimeout(resolve, 1000)})
}
```

Also, the `EventEmitter2` events are now not typed, i.e. `.on('step', (e) => {})`, now `e` is cast as `any`

This is breaking and should be resolved.

TODO

- [ ] Ensure async events work as before
- [ ] Ensure events are typed
- [ ] Add a test which ensures async events work as in the past

### PR DESCRIPTION
**Pull request by acolytec3**
Fixes #2366 

This borrows code from the old `async-eventemitter` module, migrates it to typescript, and replaces `eventemitter2` with our original API usage (hopefully).

### CODE REVIEW

**Comment by acolytec3** [comment]
@thevaizman, I re-implemented `async-eventemitter` in Typescript within our repo and slimmed it down to just overwrite the `emit` property (which is the only bit of the old `async-eventemitter` module we used internally).  When you have a chance, could you try pointing your local evm.codes build to this branch and see if everything works correctly (including your async event logic)?

**Comment by thevaizman** [comment]
@acolytec3 hey, unfortunately, I am not able to test this.
there is no easy way to simply point all packages to the git branch since this repo is a monorepo. this means that i have to build and pack using `yarn pack` each package and point to the local `.tgz` file on my machine from the `package.json`.
while this works fine, the inner dependencies are then fetched remote and i am not able to make `yarn` resolve those locally (since i can't use `yarn link` as it is incompatible with SWC 😞 )

i think i will have to wait until this change is "released" to test it...

**Comment by acolytec3** [comment]
Interesting. What were you doing before that allowed you to test against local builds of our code?

**Comment by acolytec3** [comment]
@holgerd77 provided others on the team validate this, could we do a beta release with this PR merged to verify that everything works correctly with evm.codes (and hopefully ganache too)?

**Comment by thevaizman** [comment]
> Interesting. What were you doing before that allowed you to test against local builds of our code?

changes before didn't include any inner dependencies changes (i.e `@ethereumjs/util`) afaik. therefore, i could simply use `yarn pkg` and install the local packages without problems.
however, when trying to do the same thing now, i get the following error:
```typescript
./node_modules/@ethereumjs/evm/src/evm.ts:4:3
Type error: Module '"@ethereumjs/util"' has no exported member 'AsyncEventEmitter'.

  2 | import {
  3 |   Address,
> 4 |   AsyncEventEmitter,
    |   ^
  5 |   KECCAK256_NULL,
  6 |   MAX_INTEGER,
  7 |   bigIntToBuffer,
error Command failed with exit code 1.
```

which as far as i understand stems from the inner dependency `@ethereumjs/util` being resolved remotely as part of the `evm` package. obviously, the remote `util` package is different than the local one, as it doesn't have the `AsyncEventEmitter` class, which causes this error.

File: packages/util/src/asyncEventEmitter.ts
Comment (holgerd77) [inline packages/util/src/asyncEventEmitter.ts:1]: Just so that we do not forget: please add license and a source reference here, both to the async-eventemitter package as well as the corresponding definitely-typed TS files.

**Review by gabrocheleau — APPROVED** [review]

**Review by gabrocheleau — DISMISSED** [review]

File: packages/evm/test/asyncEvents.spec.ts
Comment (jochem-brouwer) [inline packages/evm/test/asyncEvents.spec.ts:18]:
```suggestion
      t.ok(Date.now() > (startTime + 999), 'evm paused on step function for one second')
```

More explicit check

**Review by jochem-brouwer — COMMENTED** [review]
This looks great, but this is an error we did not have in the past: 

image

**Comment by gabrocheleau** [comment]
> This looks great, but this is an error we did not have in the past:
> 
> image

I thought I had addressed this in an earlier commit. Which file is that?

File: packages/evm/test/asyncEvents.spec.ts
Comment (gabrocheleau) [inline packages/evm/test/asyncEvents.spec.ts:9]:
```suggestion
tape('async events', async (t) => {
```

**Comment by jochem-brouwer** [comment]
@gabrocheleau `asyncEvents.spec.ts`

**Comment by gabrocheleau** [comment]
> @gabrocheleau `asyncEvents.spec.ts`

I don't hav ethat locally. Did you pull latest?

**Comment by jochem-brouwer** [comment]
@gabrocheleau I have and you should too; https://github.com/ethereumjs/ethereumjs-monorepo/blob/localized-asynceventemitter/packages/evm/test/asyncEvents.spec.ts

**Comment by gabrocheleau** [comment]
> @gabrocheleau I have and you should too; https://github.com/ethereumjs/ethereumjs-monorepo/blob/localized-asynceventemitter/packages/evm/test/asyncEvents.spec.ts

I see `next?.()`. Did you mean that optional chaining wasn't necessary in the past?

**Comment by jochem-brouwer** [comment]
@gabrocheleau I thought it was, but I just rechecked (`git checkout 6eb5a85a48c3b6027ccf1d59a177c4103eaca0b1` - this was the commit right before we switched to eventemitter2). It indeed raises the same error, so dismiss my comment.

**Review by jochem-brouwer — APPROVED** [review]
I have checked the behavior and the type casting, this seems to match how all worked before we went to eventemitter2, so I will approve here. Thanks @acolytec3 :+1: :smile:

**Review by gabrocheleau — APPROVED** [review]

**Comment by acolytec3** [comment]
With great fear and trembling, I merge this. 😬

@holgerd77 would it be possible to do a beta release of this so we can see if it works for evm.codes? I could try the same with ganache if I can figure out how to build their monorepo from source.

**Comment by holgerd77** [comment]
> With great fear and trembling, I merge this. 😬
> 
> @holgerd77 would it be possible to do a beta release of this so we can see if it works for evm.codes? I could try the same with ganache if I can figure out how to build their monorepo from source.

Great, Andrew (and everyone else), thanks for working through this! 

I will directly do a "normal" release. Ganache is already in a broken state if I am correct, so it can only get better from here upon. If this still do not work, we'll just do a follow-up release.

(weekend is also a reason, since I am not sure if I'll be able to work during the next 2 days)

**Pull request merged**
